// Copyright 2021 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// This file defines data structures that used to access and control SPU.
// - IO access (data infeed & outfeed).
// - Runtime control (runtime config and setup)
//
// It does NOT define how to program on SPU, programming part is defined by
// [pphlo](spu/dialect/pphlo.td)

syntax = "proto3";

package spu;

// The SPU datatype
enum DataType {
  DT_INVALID = 0;

  DT_I1 = 1;    // 1bit integer (bool).
  DT_I8 = 2;    // int8
  DT_U8 = 3;    // uint8
  DT_I16 = 4;   // int16
  DT_U16 = 5;   // uint16
  DT_I32 = 6;   // int32
  DT_U32 = 7;   // uint32
  DT_I64 = 8;   // int64
  DT_U64 = 9;   // uint64
  DT_FXP = 10;  // Fixed point type.
}

// The visibility type.
//
// SPU is secure evaluation runtime, but not all data are secret, some of them
// are publicly known to all parties, mark them as public will improve
// performance significantly.
enum Visibility {
  VIS_INVALID = 0;
  VIS_SECRET = 1;  // Invisible(unknown) for all or some of the parties.
  VIS_PUBLIC = 2;  // Visible(public) for all parties.
}

// @exclude
// The shape information.
message ShapeProto {
  repeated int64 dims = 1;
}

// Plaintext type
//
// SPU runtime does not process with plaintext directly, plaintext type is
// mainly used for IO purposes, when converting a plaintext buffer to an SPU
// buffer, we have to let spu know which type the plaintext buffer is.
enum PtType {
  PT_INVALID = 0;  //
  PT_I8 = 1;       // int8_t
  PT_U8 = 2;       // uint8_t
  PT_I16 = 3;      // int16_t
  PT_U16 = 4;      // uint16_t
  PT_I32 = 5;      // int32_t
  PT_U32 = 6;      // uint32_t
  PT_I64 = 7;      // int64_t
  PT_U64 = 8;      // uint64_t
  PT_F32 = 9;      // float
  PT_F64 = 10;     // double
  PT_I128 = 11;    // int128_t
  PT_U128 = 12;    // uint128_t
  PT_BOOL = 13;    // bool
}

// A security parameter type.
//
// The secure evaluation is based on some algebraic structure (ring or field),
enum FieldType {
  FT_INVALID = 0;
  FM32 = 1;   // Ring 2^32
  FM64 = 2;   // Ring 2^64
  FM128 = 3;  // Ring 2^128
}

// The protocol kind.
enum ProtocolKind {
  // Invalid protocol.
  PROT_INVALID = 0;

  // The reference implementation in `ring^2k`, note: this 'protocol' only
  // behave-like a fixed point secure protocol without any security guarantee.
  // Hence, it should only be selected for debugging purposes.
  REF2K = 1;

  // A semi-honest multi-party protocol. This protocol requires a trusted third
  // party to generate the offline correlated randoms. Currently, SecretFlow by
  // default ship this protocol with a trusted first party. Hence, it should
  // only be used for debugging purposes.
  SEMI2K = 2;

  // A honest majority 3PC-protocol. SecretFlow provides the semi-honest
  // implementation without Yao.
  ABY3 = 3;

  // The famous [Cheetah](https://eprint.iacr.org/2022/207) protocol, a very
  // fast 2PC protocol.
  CHEETAH = 4;
}

// The spu Value proto, used for spu value serialization.
message ValueProto {
  // The data type.
  DataType data_type = 1;

  // The data visibility.
  Visibility visibility = 2;

  // The shape of the value.
  ShapeProto shape = 3;

  // The storage type, defined by the underline evaluation engine.
  // i.e. `aby3.AShr<FM64>` means an aby3 arithmetic share in FM64.
  // usually, the application does not care about this attribute.
  string storage_type = 4;

  // The runtime/protocol dependent value data.
  bytes content = 5;
}

//////////////////////////////////////////////////////////////////////////
// Runtime configuration
//////////////////////////////////////////////////////////////////////////

// The SPU runtime configuration.
message RuntimeConfig {
  //

  ///////////////////////////////////////
  // Basic
  ///////////////////////////////////////

  // The protocol kind.
  ProtocolKind protocol = 1;

  // The field type.
  FieldType field = 2;

  // Number of fraction bits of fixed-point number.
  int64 fxp_fraction_bits = 3;

  ///////////////////////////////////////
  // Advanced
  ///////////////////////////////////////

  // @exclude
  // Runtime related, reserved for [10, 50)

  // When enabled, runtime prints verbose info of the callstack, debug purpose
  // only.
  bool enable_action_trace = 10;

  // When enabled, runtime checks runtime type infos against the
  // compile-time ones, exceptions are raised if mismatches happen. Note:
  // Runtime outputs prefer runtime type infos even when flag is on.
  bool enable_type_checker = 11;

  // When enabled, runtime prints executed pphlo list, debug purpose only.
  bool enable_pphlo_trace = 12;

  // When enabled, runtime dumps executed executables in the dump_dir, debug
  // purpose only.
  bool enable_processor_dump = 13;
  string processor_dump_dir = 14;

  // When enabled, runtime records detailed pphlo timing data, debug purpose
  // only.
  bool enable_pphlo_profile = 15;

  // When enabled, runtime records detailed hal timing data, debug purpose only.
  bool enable_hal_profile = 16;

  // Allow runtime to reveal `secret variable` use as if and while
  // condition result, debug purpose only.
  bool reveal_secret_condition = 17;

  // Allow runtime to reveal `secret variable` use as indices, debug purpose
  // only.
  bool reveal_secret_indicies = 18;

  // The public random variable generated by the runtime, the concrete prg
  // function is implementation defined.
  // Note: this seed only applies to `public variable` only, it has nothing
  // to do with security.
  uint64 public_random_seed = 19;

  // @exclude
  // Fixed-point arithmetic related, reserved for [50, 100)

  // The iterations use in f_div with Goldschmidt method.
  // 0(default) indicates implementation defined.
  int64 fxp_div_goldschmidt_iters = 50;

  // The exponential approximation method.
  enum ExpMode {
    EXP_DEFAULT = 0;  // Implementation defined.
    EXP_PADE = 1;     // The pade approximation.
    EXP_TAYLOR = 2;   // Taylor series approximation.
  }

  // The exponent approximation method.
  ExpMode fxp_exp_mode = 51;

  // Number of iterations of `exp` approximation, 0(default) indicates impl
  // defined.
  int64 fxp_exp_iters = 52;

  // The logarithm approximation method.
  enum LogMode {
    LOG_DEFAULT = 0;  // Implementation defined.
    LOG_PADE = 1;     // The pade approximation.
    LOG_NEWTON = 2;   // The newton approximation.
  }

  // The logarithm approximation method.
  LogMode fxp_log_mode = 53;

  // Number of iterations of `log` approximation, 0(default) indicates
  // impl-defined.
  int64 fxp_log_iters = 54;

  // Number of orders of `log` approximation, 0(default) indicates impl defined.
  int64 fxp_log_orders = 55;

  // The sigmoid approximation method.
  enum SigmoidMode {
    // Implementation defined.
    SIGMOID_DEFAULT = 0;
    // Minmax approximation one order.
    // f(x) = 0.5 + 0.125 * x
    SIGMOID_MM1 = 1;
    // Piece-wise simulation.
    // f(x) = 0.5 + 0.125x if -4 <= x <= 4
    //        1            if       x > 4
    //        0            if  -4 > x
    SIGMOID_SEG3 = 2;
    // The real definition, which depends on exp's accuracy.
    // f(x) = 1 / (1 + exp(-x))
    SIGMOID_REAL = 3;
  }

  // The sigmoid function approximation model.
  SigmoidMode sigmoid_mode = 56;

  /// - MPC protocol related definitions.
}

//////////////////////////////////////////////////////////////////////////
// Compiler relate definition
//////////////////////////////////////////////////////////////////////////

// Internal representation used by compiler
message XlaMeta {
  repeated Visibility inputs = 1;
}


// The executable format accepted by SPU runtime.
//
// - Inputs should be prepared before running executable.
// - Output is maintained after execution, and can be fetched by output name.
//
// ```python
//   rt = spu.Runtime(...)            # create an spu runtime.
//   rt.set_var('x', ...)             # set variable to the runtime.
//   exe = spu.ExecutableProto(       # prepare the executable.
//           name = 'balabala',
//           input_names = ['x'],
//           output_names = ['y'],
//           code = ...)
//   rt.run(exe)                      # run the execubable.
//   y = rt.get_var('y')              # get the executable from spu runtime.
// ```
message ExecutableProto {
  // The name of the executable.
  string name = 1;

  // The input names.
  repeated string input_names = 2;

  // The output names.
  repeated string output_names = 4;

  // The bytecode of the program, with format IR_MLIR_SPU.
  bytes code = 6;
}
